plugins/single_plugins/vswipe.cpp

```diff
@@ -0,0 +1,387 @@
+            {
+                start_swipe(VERTICAL);
+            }
+
+            if (state.direction == UNKNOWN)
+                return;
+        }
+
+        const double cap = speed_cap->as_cached_double();
+        const double fac = speed_factor->as_cached_double();
+
+        state.delta_prev = state.delta_last;
+        if (state.direction == HORIZONTAL)
+        {
+            state.delta_sum += vswipe_process_delta(ev->dx, state.delta_sum,
+                state.vx, state.vw, cap, fac);
+            state.delta_last = ev->dx;
+        } else
+        {
+            state.delta_sum += vswipe_process_delta(ev->dy, state.delta_sum,
+                state.vy, state.vh, cap, fac);
+            state.delta_last = ev->dy;
+        }
+
+        output->render->damage_whole();
+    };
+
+    wf::signal_callback_t on_swipe_end = [=] (wf::signal_data_t *data)
+    {
+        if (!state.swiping)
+            return;
+
+        state.swiping = false;
+
+        auto ev = static_cast<wf::swipe_end_signal*>(data)->ev;
+
+        const double move_threshold =
+            clamp(threshold->as_cached_double(), 0.0, 1.0);
+        const double fast_threshold =
+            clamp(delta_threshold->as_cached_double(), 0.0, 1000.0);
+
+        int target_delta = 0;
+        wf_point target_workspace = {state.vx, state.vy};
+
+        if (!ev->cancelled || ignore_cancel->as_int())
+        {
+            switch (state.direction)
+            {
+                case UNKNOWN:
+                    target_delta = 0;
+                    break;
+                case HORIZONTAL:
+                    target_delta = vswipe_finish_target(state.delta_sum,
+                        state.vx, state.vw, state.delta_prev + state.delta_last,
+                        move_threshold, fast_threshold);
+                    target_workspace.x -= target_delta;
+                    break;
+                case VERTICAL:
+                    target_delta = vswipe_finish_target(state.delta_sum,
+                        state.vy, state.vh, state.delta_prev + state.delta_last,
+                        move_threshold, fast_threshold);
+                    target_workspace.y -= target_delta;
+                    break;
+            }
+        }
+
+        transition = {state.delta_sum, target_delta + state.gap * target_delta};
+
+        output->workspace->set_workspace(target_workspace);
+        output->render->set_redraw_always();
+        duration.start();
+    };
+
+    void finalize_and_exit()
+    {
+        state.swiping = false;
+        grab_interface->ungrab();
+        output->deactivate_plugin(grab_interface);
+
+        if (streams.prev.running)
+            output->render->workspace_stream_stop(streams.prev);
+        output->render->workspace_stream_stop(streams.curr);
+        if (streams.next.running)
+            output->render->workspace_stream_stop(streams.next);
+
+        output->render->set_renderer(nullptr);
+        output->render->set_redraw_always(false);
```

Comment:
It seems that the `set_redraw_always(true)` won't be always called, right? So the the `set_redraw_always(false)` shouldn't be called always too, otherwise we may corrupt the counter.

Reply:
there is a counter?! hmm okay

Reply:
yes, if you call set_redraw_always() it increases the counter ....